include/eeprom/core/accessor.hpp

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include <array>
+#include <vector>
+
+#include "addresses.hpp"
+#include "task.hpp"
+#include "types.hpp"
+
+namespace eeprom {
+namespace accessor {
+
+/**
+ * Interface of an object that listens for reads
+ */
+template <typename T>
+class ReadListener {
+  public:
+    ReadListener() = default;
+    ReadListener(const ReadListener&) = default;
+    auto operator=(const ReadListener&) -> ReadListener& = default;
+    ReadListener(ReadListener&&) noexcept = default;
+    auto operator=(ReadListener&&) noexcept -> ReadListener& = default;
+
+    virtual ~ReadListener() = default;
+    virtual void on_read(const T&) = 0;
+};
+
+template <task::TaskClient EEPromTaskClient, typename T,
+          types::address data_begin, types::address default_size>
+class EEPromAccessor {
+  public:
+    explicit EEPromAccessor(EEPromTaskClient& eeprom_client,
+                            ReadListener<T>& listener)
+        : eeprom_client{eeprom_client},
+          read_listener{listener},
+          begin{data_begin} {}
+    EEPromAccessor(const EEPromAccessor&) = delete;
+    EEPromAccessor(EEPromAccessor&&) = delete;
+    auto operator=(EEPromAccessor&&) -> EEPromAccessor& = delete;
+    auto operator=(const EEPromAccessor&) -> EEPromAccessor& = delete;
+    virtual ~EEPromAccessor() = default;
+    /**
+     * Begin a read of the data
+     */
+    auto start_read() -> void { start_read_at_offset(0, default_size); }
+
+    /**
+     * Write data to eeprom
+     */
+    auto write(const T& type_value) -> void {
+        write_at_offset(type_value, 0, default_size);
+    }
+
+  protected:
+    T type_data = T(default_size);
+    EEPromTaskClient& eeprom_client;
+    ReadListener<T>& read_listener;
+    types::address begin;
+
+    /**
+     * Write data to eeprom at a specified offset
+     */
+    auto write_at_offset(const std::vector<uint8_t>& data,
```

Comment:
factoring this into taking an iterator pair or a range means this would not need a direct reference to `vector` anymore